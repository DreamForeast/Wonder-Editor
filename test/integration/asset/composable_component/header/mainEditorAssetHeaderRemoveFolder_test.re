open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open Js.Promise;

let _ =
  describe("MainEditorAssetHeader->remove folder", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initState(~sandbox, ());

      InspectorCanvasTool.prepareInspectorEngineState(sandbox);

      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });

    afterEach(() => {
      restoreSandbox(refJsObjToSandbox(sandbox^));
      StateEditorService.getState()
      |> CurrentNodeIdAssetEditorService.clearCurrentNodeId
      |> SelectedFolderNodeIdInAssetTreeAssetEditorService.clearSelectedFolderNodeIdInAssetTree
      |> StateEditorService.setState
      |> ignore;
    });

    test("click remove-button should remove folder from assetTreeRoot", () => {
      let assetTreeData =
        MainEditorAssetTreeTool.BuildAssetTree.Folder.TwoLayer.buildOneFolderAssetTree();
      let component = BuildComponentTool.buildAssetComponent();

      MainEditorAssetHeaderOperateNodeTool.removeFolderNode(
        ~folderNodeId=
          MainEditorAssetTreeTool.BuildAssetTree.Folder.TwoLayer.getFirstFolderNodeId(
            assetTreeData,
          ),
        (),
      );

      BuildComponentTool.buildAssetComponent()
      |> ReactTestTool.createSnapshotAndMatch;
    });

    describe("should remove folder's all children", () => {
      beforeEach(() => {
        MainEditorSceneTool.initStateWithJob(
          ~sandbox,
          ~isBuildFakeDom=false,
          ~noWorkerJobRecord=
            NoWorkerJobConfigToolEngine.buildNoWorkerJobConfig(
              ~loopPipelines=
                {|
                   [
                       {
                           "name": "default",
                           "jobs": [
                               {
                                   "name": "dispose"
                               }
                           ]
                       }
                   ]
               |},
              (),
            ),
          (),
        );
        MainEditorSceneTool.prepareScene(sandbox);
      });

      test("test remove material asset child", () => {
        open NodeAssetType;

        let addedFolderNodeId = MainEditorAssetIdTool.getNewAssetId();

        MainEditorAssetHeaderOperateNodeTool.addFolder();

        MainEditorAssetTreeTool.Select.selectFolderNode(
          ~nodeId=addedFolderNodeId,
          (),
        );

        let materialNodeId = MainEditorAssetIdTool.getNewAssetId();
        MainEditorAssetHeaderOperateNodeTool.addMaterial();

        let {snapshotImageDataIndex} =
          StateEditorService.getState()
          |> OperateTreeAssetEditorService.unsafeFindNodeById(materialNodeId)
          |> MaterialNodeAssetService.getNodeData;

        MainEditorAssetHeaderOperateNodeTool.removeFolderNode(
          ~folderNodeId=addedFolderNodeId,
          (),
        );

        StateEditorService.getState()
        |> BasicSourceTextureImageDataMapAssetEditorService.getData(snapshotImageDataIndex)
        |> Js.Option.isNone
        |> expect == true;
      });
      describe("test remove texture asset child", () => {
        beforeEach(() => {
          MainEditorAssetTool.buildFakeImage();
          MainEditorAssetTool.buildFakeFileReader();
        });

        testPromise("test remove it in child folder", () => {
          let addedFolderNodeId1 = MainEditorAssetIdTool.getNewAssetId();

          MainEditorAssetHeaderOperateNodeTool.addFolder();

          MainEditorAssetTreeTool.Select.selectFolderNode(
            ~nodeId=addedFolderNodeId1,
            (),
          );

          let addedFolderNodeId2 = MainEditorAssetIdTool.getNewAssetId();

          MainEditorAssetHeaderOperateNodeTool.addFolder();

          MainEditorAssetTreeTool.Select.selectFolderNode(
            ~nodeId=addedFolderNodeId2,
            (),
          );

          MainEditorAssetUploadTool.loadOneTexture()
          |> Js.Promise.then_(uploadedTextureNodeId1 => {
               let textureComponent =
                 MainEditorAssetTextureNodeTool.getTextureComponent(
                   uploadedTextureNodeId1,
                 )
                 |> StateLogicService.getEditorState;

               MainEditorAssetHeaderOperateNodeTool.removeFolderNode(
                 ~folderNodeId=addedFolderNodeId1,
                 (),
               );

               BasicSourceTextureToolEngine.isAlive(textureComponent)
               |> StateLogicService.getEngineStateToGetData
               |> expect == false
               |> resolve;
             });
        });
      });
    });
  });