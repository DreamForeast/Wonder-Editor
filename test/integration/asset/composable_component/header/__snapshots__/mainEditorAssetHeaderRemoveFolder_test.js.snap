// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorAssetHeader->remove folder click remove-button should remove folder from assetTreeRoot 1`] = `
<article
  className="wonder-asset-component"
>
  <div
    className="asset-tree"
  >
    <article
      className="wonder-asset-header"
    >
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="add asset"
      >
        <div
          className="item-canBeClick"
        >
          <img
            src="./public/img/add.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
        onClick={[Function]}
        title="remove asset"
      >
        <div
          className="item-notBeClick"
        >
          <img
            src="./public/img/notRemove.png"
          />
        </div>
      </div>
      <div
        className="asset-header-item"
      >
        <div
          className="item-canBeClick"
          title="upload asset"
        >
          <img
            src="./public/img/load.png"
          />
          <input
            className="asset-fileLoad"
            multiple={false}
            onChange={[Function]}
            type="file"
          />
        </div>
      </div>
    </article>
    <article
      className="wonder-asset-assetTree"
    >
      <ul
        className="wonder-tree-node"
      >
        <li>
          <div
            className="item-triangle"
          />
          <img
            className="treeNode-icon"
            src="./public/img/package.png"
          />
          <div
            className="draggable-container select-not-active"
            draggable={true}
            onClick={[Function]}
            onDragEnd={[Function]}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            style={Object {}}
          >
            Assets
          </div>
        </li>
      </ul>
    </article>
  </div>
  <article
    className="wonder-asset-assetChildren"
  />
</article>
`;
